### ISSUE TITLE
random.split() should allow for passing in a desired shape

### ISSUE THREAD

**Issue by shoyer** [google/jax#4013]
The current API is `random.split(key, num)`, which returns an array of shape `(num, 2)`.

I'd like to have an API of the form `random.split(key, shape)`, which returns an array of shape `shape + (2,)`, without requiring an additional `reshape`. This would make `random.split` easier to use for preparing multi-dimensional key, e.g., for use with repeated `vmap`.

Some options:

1. Write a function with a new name for this API, e.g., `random.split_with_shape` (this is a bad name)
2. Overload the `num` argument to allow for _either_ an integer size or a tuple of integers `shape`. This is a little tricky to get right, but in principle could be done (NumPy has lot of APIs like this).
3. Make `num` optional and add a new optional `shape` parameter that could be used instead.
4. Deprecate `num` and replace it with `shape`. This would involve (2) and an extra temporary keyword-only argument. Eventually, we could require either (a) the new argument name `shape` or (b) the new argument name `shape` _and_ passing a tuple of integers instead of an integer.

I would lean towards option (3). It's a little messy to add an extra argument but it is cleaner than overloading `random.split` to support either sizes or shapes. The later would be consistent with how `numpy.random` works with the `size` argument, but not how the rest of `jax.random` works.

**Comment by jakevdp** [comment]
I like option 3 as well, though 2 also appeals to me.

Offhand, do you know why JAX tends not to allow integers in places where shapes are expected? That's something that I've found pretty jarring moving to jax from numpy. Was this a deliberate decision?

**Comment by shoyer** [comment]
> Offhand, do you know why JAX tends not to allow integers in places where shapes are expected? That's something that I've found pretty jarring moving to jax from numpy. Was this a deliberate decision?

I don't know who wrote the first version of `random.py`. I think we'd have to ask @mattjj on this one.

**Comment by jakevdp** [comment]
Assigning to @mattjj for design opinions about polymorphic `num`/`shape` vs. adding a new shape parameter

**Comment by mattjj** [comment]
I wrote the first version! I'll take a look at this when I can.

**Comment by minedoc** [comment]
Hello, would it be possible to add something like this:

```py
def split_shape(key, shape):
  if type(shape) in (tuple, list):
    return tuple(split_shape(k, s) for (k, s) in zip(jax.random.split(key, len(shape)), shape))
  elif shape == 0:
    return key
  else:
    return jax.random.split(key, shape)
```

For example:

```py
split_shape(k, 4) = split(k, 4)
split_shape(k, [4,4]) = (split(k1, 4), split(k2, 4))
```

**Comment by jakevdp** [comment]
I would implement it this way:
```python
def split_shape(key, shape):
  size = math.prod(shape) if isinstance(shape, (tuple, list)) else shape
  return jax.random.split(key, size).reshape(*shape, *key.shape)
```
Whether this logic should be contained in `random.split()` itself, I'm not sure.

**Comment by minedoc** [comment]
The thing I was looking for, maybe it's not common, was to have recursively defined shape:

```py
a, bs = split_shape(k, (0, [2]*5))

model += linear(a)
for b0, b1 in bs:
    model += linear(b0)
    model += linear(b1)
```

**Comment by carlosgmartin** [comment]
Could the type annotation of `num` be amended to allow lists?

Perhaps it should be annotated with `Shape` instead of `tuple[int, ...]`.

**Comment by NeilGirdhar** [comment]
Isn't `Shape` already `tuple[int, ....]`?

**Comment by carlosgmartin** [comment]
@NeilGirdhar According to [this line](https://github.com/google/jax/blob/main/jax/_src/random.py#L56), `Shape` is `Sequence[int]`.

**Comment by NeilGirdhar** [comment]
I see.  Lots of other `Shape` definitions in Jax are tuples and shape is a tuple in the Array API too.   I'm not sure what the Jax team will decide, but I personally think Jax should prefer tuple in general.  Narrow annotations catch more mistakes.

**Comment by jakevdp** [comment]
> @NeilGirdhar According to [this line](https://github.com/google/jax/blob/main/jax/_src/random.py#L56), `Shape` is `Sequence[int]`.

The reason for this is that JAX was originally written without strict type contracts for such APIs. When we began adding type annotations we found that existing usage used general sequences for shapes, and we annotated the APIs to describe their existing behavior. I tend to agree that tuples make more sense for shapes, and so when adding new shape arguments to APIs, we've annotated them more narrowly.
